### PR TITLE
Move breadcrumbs setting from views to controllers

### LINKED ISSUE #3149: Breadcrumbs displays invalid emails modifying users
When changing a user's email in the backend (`/admin/users/:id`), if there is a validation error on the email, breadcrumbs of the page are updated with the wrong, invalid value.

**Solidus Version**:

master

**To Reproduce**

In sandbox:

1. Go to Admin -> User -> Any user -> Account tab
2. Change user's email with the email taken by another user and save
3. Click Update button

**Current behavior**

Breadcrumbs are updated with the wrong email, even if there's an error on the page

**Expected behavior**

The old, persisted and valid email is displayed in breadcrumbs

**Screenshots**

breadcrumbs-bug

👆 I'm actually editing `admin@example.com`

**Additional context**

I suppose this happens because breacrumbs is using the ivar (maybe @user, or @object from [`Spree::Admin::ResourceController`](https://github.com/solidusio/solidus/blob/14a9eb15eb22c92ef5e00e4daa74f96e58ced7b9/backend/app/controllers/spree/admin/resource_controller.rb)) that contains the instance of the User with errors (and previously submitted params) and not the persisted one, which is valid.

I'm afraid this could happen in other pages as well, where the value that the user is trying to change to an invalid one is used to create breadcrumbs of that page.

### PR DESCRIPTION
**Pull request by cavalle**
**Description**

This PR moves the definition of all the breadcrumbs in the admin module from the views to a `before_action` method in the controllers. This fixes #3149 and other similar bugs, and will avoid reintroducing them in the future.

Since this change affects all the breadcrumbs in the admin engine, the PR is quite large and so I've split it in several cohesive commits. The description of each commit elaborates on the changes implemented in it. I suggest reviewing this PR commit by commit.

These are the key decisions I made:

1. This PR implements the general approach suggested by @kennyadsl in #3152. That is, by setting the breadcrumbs in a `before_action` of the controller, we ensure the logic of the actions haven't changed yet any objects used in the breadcrumb, thus avoiding bugs like solidusio/solidus#3149.

2. I took a look at https://github.com/weppos/breadcrumbs_on_rails and (the seemingly better maintained) https://github.com/piotrmurach/loaf for inspiration, but finally decided to go for a simpler solution based on the existing helpers.

3. After these changes, every controller in the admin module (that renders breadcrumbs) includes a `set_breadcrumbs` method that is called for all the actions of the controller. Since different actions usually have different breadcrumbs, after the common breadcrumbs are set, this method usually uses a few conditionals to set the action-specific breadcrumbs 

    An alternative solution would be to set the common breadcrumbs in a common method while any action-specific breadcrumb at the top of each action method. I chose to put everything in the same method because I like better to have everything (both shared and action-specific breadcrumbs) in the same place, rather than scattered around the controller. However, I don't have a very strong opinion about this choice, I understand the merits of the aforementioned alternative.

You can find more details about the changes in each commit's description.

**Checklist:**
- [ ] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [ ] I have added a detailed description into each commit message
- [ ] I have added tests to cover this change

### CODE REVIEW

File: backend/app/views/spree/admin/return_authorizations/_return_authorization_breadcrumb.html.erb [hunk 1]
```diff
@@ -0,0 +1 @@
+<%= text %> (<%= t(state, scope: 'spree.return_authorization_states')%>)
```
[the remark this answers is not in the document]
Comment (cavalle) [reply backend/app/views/spree/admin/return_authorizations/_return_authorization_breadcrumb.html.erb]: Oops! I only checked with Rubocop locally. I'll fix these!

**Comment by cavalle** [comment]
@kennyadsl thanks for your very clear feedback! I'll respond and work on it during this morning.

File: backend/app/controllers/spree/admin/adjustment_reasons_controller.rb [hunk 2]
```diff
@@ -3,6 +3,17 @@
...
     class AdjustmentReasonsController < ResourceController
+      before_action :set_breadcrumbs
+
+      private
+
+      def set_breadcrumbs
+        add_breadcrumb t('spree.settings')
+        add_breadcrumb t('spree.admin.tab.checkout')
+        add_breadcrumb plural_resource_name(Spree::AdjustmentReason), spree.admin_adjustment_reasons_path
+        add_breadcrumb @adjustment_reason.name          if params[:id].present?
```
[the remark this answers is not in the document]
Comment (cavalle) [reply backend/app/controllers/spree/admin/adjustment_reasons_controller.rb]: Thanks for letting me know about this convention. I wasn't aware of it. I'll update everywhere I'm using spaces to align code.

File: backend/app/controllers/spree/admin/base_controller.rb [hunk 3]
```diff
@@ -62,6 +62,28 @@ def lock_order
       def order_mutex_redirect_path
         edit_admin_order_path(@order)
       end
+
+      def add_breadcrumb(name, path = nil, render_options = {})
```
[the remark this answers is not in the document]
Comment (cavalle) [reply backend/app/controllers/spree/admin/base_controller.rb]: I'm glad that you make this suggestion because I agree a concern here is a nice way to keep code organised. The reason I didn't use one is that I found that the `backend` gem doesn't use concerns at all for any other purpose so I wasn't sure using one here would be well received.

I'll move these into a concern right away!

File: backend/app/views/spree/admin/shared/_breadcrumb_item.html.erb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<% if path.present? && !last %>
+<%= link_to text, path %>
```
[the remark this answers is not in the document]
Comment (cavalle) [reply backend/app/views/spree/admin/shared/_breadcrumb_item.html.erb]: Yes, there was a reason: indenting those two lines renders extra spaces around the text that, for some reason, cause the text to appear misplaced. I'll take another quick look to see if I can understand why the those spaces aren't ignored by the browser.

File: backend/spec/features/admin/users_spec.rb [hunk 5]
```diff
@@ -242,6 +242,17 @@ def always_invalid_email
           expect(page).to have_content("Email is invalid")
         end
       end
+
+      # Regression test for https://github.com/solidusio/solidus/issues/3149
+      it 'should show persisted email in breadcrumb' do
```
[the remark this answers is not in the document]
Comment (cavalle) [reply backend/spec/features/admin/users_spec.rb]: I'll update everything to that convention!

File: backend/spec/helpers/admin/navigation_helper_spec.rb [hunk 6]
```diff
@@ -136,4 +136,31 @@
...
+    it "should render breadcrumb with link when it isn't the last one" do
+      expect(subject).to have_link("first", href: "http://first.example.com")
+    end
+
+    it "should render breadcrumb without link when it's the last one" do
+      expect(subject).to_not have_link("fourth")
+      expect(subject).to have_content("fourth")
+    end
+
+    it "should render breadcrumb without link when path isn't given" do
```
[the remark this answers is not in the document]
Comment (cavalle) [reply backend/spec/helpers/admin/navigation_helper_spec.rb]: Same response ;)

File: core/spec/lib/spree/core/controller_helpers/common_spec.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    let(:plural_config) { Spree::I18N_GENERIC_PLURAL }
+    let(:base_class) { Spree::Product }
+
+    subject { controller.plural_resource_name base_class }
+
+    it "should use ActiveModel::Naming module to pluralize model names" do
+      expect(subject).to eq base_class.model_name.human(count: plural_config)
+    end
+
+    it "should use the Spree::I18N_GENERIC_PLURAL constant" do
```
[the remark this answers is not in the document]
Comment (cavalle) [reply core/spec/lib/spree/core/controller_helpers/common_spec.rb]: Same response ;)

File: backend/app/views/spree/admin/return_authorizations/_return_authorization_breadcrumb.html.erb [hunk 8]
```diff
@@ -0,0 +1 @@
+<%= text %> (<%= t(state, scope: 'spree.return_authorization_states') %>)
```
[the remark this answers is not in the document]
Comment (cavalle) [reply backend/app/views/spree/admin/return_authorizations/_return_authorization_breadcrumb.html.erb]: This is an interesting one. This particular breadcrumb could be perfectly specified from the controller itself using simple string interpolation. However,  In fact, I did it like that in an initial version. However, the string interpolated ended up being a bit too long and a bit too complex and I, since I had to implement the ability to use custom partials for another breadcrumb, I thought it would make sense to use it in this case for clarity.

Now, one thing I wondered is whether this additional level of complexity (i.e. the ability to specify custom breadcrumbs with partials) was really justified if we only needed for one or two breadcrumbs. Everything would be simpler if all the breadcrumbs would be just text, and didn't need special HTML formatting like in [this particular case][1] in the whole app. I'd appreciate your thoughts on this. Is it really needed and valuable to show the status information with colours in the breadcrumbs?

[1]:https://github.com/solidusio/solidus/pull/3191/files#diff-e249441f4fa5641286d474939a6db2f3

File: backend/app/views/spree/admin/payments/credit.html.erb [hunk 9]
```diff
@@ -1,18 +0,0 @@
-<%= render partial: 'spree/admin/shared/order_tabs', locals: { current: 'Creditcards' } %>
```
[the remark this answers is not in the document]
Comment (cavalle) [reply backend/app/views/spree/admin/payments/credit.html.erb]: Right! I gave details about this removal in the commit (0192bc00cc41099aa650c1b6818514686e3e24cd) but I forgot to stress in the PR description that I needed you to confirm we could to this. 

While moving the breadcrumb logic to controllers, I found that this template didn't seem to be used. Basically, I couldn't find any route that would use it or any other code referencing it. Also, no test broke after removing this template. But, of course, you know the codebase better than me, and you need to confirm it's ok to remove it.

**Comment by cavalle** [comment]
Just pushed additional changes to address the points discussed above. I didn't rebase or amend the previous commits, but if you usually work that way let me know and I'll do it.

File: backend/app/views/spree/admin/return_authorizations/_return_authorization_breadcrumb.html.erb (hunk 8 above)
[the remark this answers is not in the document]
Comment (kennyadsl) [reply backend/app/views/spree/admin/return_authorizations/_return_authorization_breadcrumb.html.erb]: I think we can remove the HTML from breadcrumbs, even if it probably requires a new PR and an alternative UI method to show the state of the parent object in the page, which I think was the original reason why they have been introduced.

File: backend/app/views/spree/admin/payments/credit.html.erb (hunk 9 above)
[the remark this answers is not in the document]
Comment (kennyadsl) [reply backend/app/views/spree/admin/payments/credit.html.erb]: Yes, I did read that commit message (thanks for being so detailed there). My face was because I didn't expect to have an unused file in the application!

**Comment by kennyadsl** [comment]
Thanks @cavalle, I'll talk in the next core team meeting about:

- better to use a single method or define the part of the breadcrumb into each specific action?
- can we remove the status from the breadcrumb?

I hope to come back soon with a response.

**Comment by kennyadsl** [comment]
@cavalle we had a quick chat about this PR in the last Core Team meeting and this is what we decided:

> better to use a single method or define the part of the breadcrumb into each specific action?

We are not sure but the current implementation is a great starting point that we can revisit later if needed. We also discussed the possibility to create a "DSL" around this, that allows setting specific features only for specific actions, maybe something like:

```ruby
add_breadcrumb @adjustment_reason.name, only: [:show, :edit, :update]
```

That would actually take care of pushing stuff in the right action but for now, this is just an idea, what we have is great.

<hr>

> can we remove the status from the breadcrumb?

Yes, we can. We prefer to have this PR simplified than having a `pill` component into breadcrumbs. It could be nice to have but I think we should live without it or even better, move the status `pill` into another ad-hoc UI component in the page, which we can probably discuss into a separate issue.

<hr>

So missing TODO items are:

- remove status pills from breadcrumbs
- remove complex templates that did come from pills usage
- rebase against master squashing last commits to have a clean git history

Thanks a lot again for this great contribution!

**Comment by cavalle** [comment]
Awesome! I'll take care of the TODO items as soon as I can.

**Comment by cavalle** [comment]
I implemented all points in the [previous comment](https://github.com/solidusio/solidus/pull/3191#issuecomment-490814166) and CI passed. This one is ready for review again!

**Comment by kennyadsl** [comment]
@cavalle Hey Luismi, thanks! I just noticed another thing that we should take care of before merging: 

if we remove the `admin_breadcrumb` helper method at all we will break existing stores and extensions that could be using that method when creating new pages in the admin. We should instead keep it working and print a deprecation message so they can gradually transition to the new way of setting breadcrumbs. If you don't have time, don't worry, just let me know and I can take care of it in another PR asking for your review when done. 

Thanks again 🙏

**Comment by cavalle** [comment]
@kennyadsl right! that makes a lot of sense, and I'd love to take care of it, but I'm afraid I no longer have much available time. So, if you don't mind taking care of it in another PR, I'd very much appreciate it!

**Comment by kennyadsl** [comment]
Sure @cavalle, I'm thinking I could also open a PR on your fork to address the issue so my changes will be part of this one. Does it work for you?

**Comment by cavalle** [comment]
No problem with that! I made you collaborator of my fork.

**Comment by kennyadsl** [comment]
@cavalle I think I've done, the only commit I changed is https://github.com/solidusio/solidus/pull/3191/commits/5c378b5875af529b32c88ba384447e4b2429085a. I've also changed the commit message and description to reflect/explain the changes I've made. Please, let me know your thoughts if you have time.


<hr>

**Notes to add the CHANGELOG**:

Setting breadcrumbs from views has been deprecated and now should do that from the controller. 
For example, if you are setting a breadcrumb in your `spree/admin/custom_models/index.html.erb` view like this:

```erb
<% admin_breadcrumb(link_to plural_resource_name(Spree::CustomModel), spree.admin_custom_models_path) %>
```

You should remove it and add the following into your `app/controllers/spree/custom_models_controller.rb` controller:

```ruby
before_action :set_breadcrumbs, only: :index

# ...

private

# ...

def set_breadcrumbs
  add_breadcrumb plural_resource_name(Spree::CustomModel), spree.admin_custom_models_path
end

**Comment by cavalle** [comment]
@kennyadsl the changes you made look good to me :+1:

**Comment by kennyadsl** [comment]
@ericsaupe I'll take care of adding this notes to the CHANGELOG when we'll release Solidus 2.9

File: backend/app/controllers/concerns/spree/admin/breadcrumbs.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+module Spree
+  module Admin
+    module Breadcrumbs
+      def add_breadcrumb(name, path = nil)
```
Comment (tvdeyen) [inline backend/app/controllers/concerns/spree/admin/breadcrumbs.rb]: Can we make this methods `protected` so they don't become actions?

File: backend/app/controllers/concerns/spree/admin/breadcrumbs.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        @admin_breadcrumbs ||= []
```
Comment (tvdeyen) [inline backend/app/controllers/concerns/spree/admin/breadcrumbs.rb]: Can we make this a `Set`, so that we do not get potential duplicates and gain some performance?

File: backend/app/controllers/spree/admin/base_controller.rb [hunk 12]
```diff
@@ -3,6 +3,8 @@
 module Spree
   module Admin
     class BaseController < Spree::BaseController
+      include Spree::Admin::Breadcrumbs
+
```
Comment (tvdeyen) [inline backend/app/controllers/spree/admin/base_controller.rb]: Could we add the `before_action :set_breadcumbs` here and add an empty method that gets called if the child class does not implement `set_breadcrumbs`? We might even have a "defaut breadcrumb" (build from the resource) for controllers. We could remove a lot of code then.

File: backend/app/controllers/spree/admin/customer_returns_controller.rb [hunk 13]
```diff
@@ -9,6 +9,7 @@ class CustomerReturnsController < ResourceController
       before_action :parent # ensure order gets loaded to support our pseudo parent-child relationship
       before_action :load_form_data, only: [:new, :edit]
       before_action :build_return_items_from_params, only: [:create]
+      before_action :set_breadcrumbs
```
Comment (tvdeyen) [inline backend/app/controllers/spree/admin/customer_returns_controller.rb]: We could move this `before_action` into the `spree/admin/base_controller`

File: backend/app/controllers/spree/admin/orders_controller.rb [hunk 14]
```diff
@@ -186,6 +187,13 @@ def require_ship_address
           redirect_to edit_admin_order_customer_url(@order)
         end
       end
+
+      def set_breadcrumbs
+        set_order_breadcrumbs
+        add_breadcrumb t('spree.cart') if action_name == 'cart'
```
Comment (tvdeyen) [inline backend/app/controllers/spree/admin/orders_controller.rb]: We could move the breadcrumb into the `cart` action instead of this `if`

File: backend/app/controllers/spree/admin/orders_controller.rb [hunk 15]
```diff
@@ -186,6 +187,13 @@ def require_ship_address
...
+        add_breadcrumb t('spree.confirm_order') if action_name == 'confirm'
```
Comment (tvdeyen) [inline backend/app/controllers/spree/admin/orders_controller.rb]: Same here, why not add the breadcrumb only in the `confirm` action?

File: core/lib/spree/core/controller_helpers/common.rb [hunk 16]
```diff
@@ -11,12 +11,17 @@ module Common
...
           helper_method :title=
           helper_method :accurate_title
+          helper_method :plural_resource_name
 
           layout :get_layout
 
           before_action :set_user_language
         end
 
+        def plural_resource_name(resource_class)
```
Comment (tvdeyen) [inline core/lib/spree/core/controller_helpers/common.rb]: Can we move this method into the `admin/resource_controller` instead? This module gets included in many (even frontend) controllers.

**Comment by kennyadsl** [comment]
@tvdeyen Thanks for the review. I think that the edge case reported in #3149 is actually present in all admin edit forms, where "the thing" (name, title, whatever) that is rendered in the breadcrumbs has some validation. 

I see your point that this is a somehow drastic change and potentially could break something, even if I think the only issue will be with extensions The solution should use both approaches, until v2.8 goes EOL (if we merge it now), conditionally switching from a way of setting breadcrumbs to another depending on the version. Eg. in extensions:

```erb
<% if Solidus.version < '2.9' %>
  <% admin_breadcrumb(link_to plural_resource_name(Spree::CustomModel), spree.admin_custom_models_path) %>
<% end %>
```

and in the controller:

```ruby
def set_breadcrumbs
  return if Solidus.version < '2.9'

  # ... rest of the code
```

which is not ideal at all. 

**TL;DR** Despite I proposed this path, I'm not sure as well that this is the best solution. Maybe we should go with a third option? At RailsConf @joelhawksley presented [this talk](https://www.youtube.com/watch?v=y5Z5a6QdA-M) about adding a "component" layer into Rails, that is used at GitHub. Here's a [demo app](https://github.com/joelhawksley/actionview-component-demo). IMO it is a fresh approach that we could consider, even because it will be probably be merged into Rails soon.

**Comment by kennyadsl** [comment]
Closing, we'll keep this as reference for some future work. Thanks to anyone involved!
